sky/clouds/kubernetes.py

```diff
@@ -0,0 +1,364 @@
+        return name
+
+    @staticmethod
+    def is_valid_instance_type(name: str) -> bool:
+        """Returns whether the given name is a valid instance type."""
+        pattern = re.compile(r'^(\d+(\.\d+)?CPU--\d+(\.\d+)?GB)(--\d+\S+)?$')
+        return bool(pattern.match(name))
+
+    @classmethod
+    def _parse_instance_type(
+            cls,
+            name: str) -> Tuple[float, float, Optional[float], Optional[str]]:
+        """Returns the cpus, memory, accelerator_count, and accelerator_type
+        from the given name."""
+        pattern = re.compile(
+            r'^(?P<cpus>\d+(\.\d+)?)CPU--(?P<memory>\d+(\.\d+)?)GB(?:--(?P<accelerator_count>\d+)(?P<accelerator_type>\S+))?$'  # pylint: disable=line-too-long
+        )
+        match = pattern.match(name)
+        if match:
+            cpus = float(match.group('cpus'))
+            memory = float(match.group('memory'))
+            accelerator_count = match.group('accelerator_count')
+            accelerator_type = match.group('accelerator_type')
+            if accelerator_count:
+                accelerator_count = float(accelerator_count)
+                accelerator_type = str(accelerator_type)
+            else:
+                accelerator_count = None
+                accelerator_type = None
+            return cpus, memory, accelerator_count, accelerator_type
+        else:
+            raise ValueError(f'Invalid instance name: {name}')
+
+    @classmethod
+    def from_instance_type(cls, name: str) -> 'KubernetesInstanceType':
+        """Returns an instance name object from the given name."""
+        if not cls.is_valid_instance_type(name):
+            raise ValueError(f'Invalid instance name: {name}')
+        cpus, memory, accelerator_count, accelerator_type = \
+            cls._parse_instance_type(name)
+        return cls(cpus=cpus,
+                   memory=memory,
+                   accelerator_count=accelerator_count,
+                   accelerator_type=accelerator_type)
+
+    @classmethod
+    def from_resources(cls,
+                       cpus: float,
+                       memory: float,
+                       accelerator_count: float = 0,
+                       accelerator_type: str = '') -> 'KubernetesInstanceType':
+        """Returns an instance name object from the given resources."""
+        name = f'{cpus}CPU--{memory}GB'
+        if accelerator_count > 0:
+            name += f'--{accelerator_count}{accelerator_type}'
+        return cls(cpus=cpus,
+                   memory=memory,
+                   accelerator_count=accelerator_count,
+                   accelerator_type=accelerator_type)
+
+    def __str__(self):
+        return self.name
+
+
+@clouds.CLOUD_REGISTRY.register
+class Kubernetes(clouds.Cloud):
+    """Kubernetes."""
+
+    SKY_SSH_KEY_SECRET_NAME = f'sky-ssh-{common_utils.get_user_hash()}'
+
+    # TODO(romilb): Make the timeout configurable.
+    TIMEOUT = 60  # Timeout for resource provisioning
```

Comment:
TL;DR - good catch, I've reduced it to 10s (tested with `sky local up` and GKE non-autoscaling cluster).

----
The reason we have this timeout is to 1) give time to Kubernetes to schedule the pods and 2) support autoscaling in the future. Since Kubernetes keeps pods indefinitely pending till resources are available AND we don't have visibility into any underlying autoscaling, we add a timeout to give the underlying autoscaler enough time to add resources. In the future, this timeout will be configurable by the user depending on the following:

* For non-autoscaling clusters - this timeout should be close to zero, since if the pod doesn't get scheduled immediately, we know for sure there are not resources currently available.

* For autoscaling clusters - this timeout would need to be set to roughly the amount of time it takes to add a new node.

Previously, this timeout also included the time taken to pull the image so I had set it conservatively to 60s. We now break this timeout check loop the moment a pod starts pulling an image (since image pull happens only after pod has been scheduled). Thus, this can safely be reduced to only account for scheduling time, and I found that 10s is a safe timeout to use.